base/reinterpretarray.jl

```diff
@@ -0,0 +1,127 @@
+"""
+Gives a reinterpreted view (of element type T) of the underlying array (of element type S).
+If the size of `T` differs from the size of `S`, the array will be compressed/expanded in
+the first dimension.
+"""
+struct ReinterpretArray{T,N,S,A<:AbstractArray{S, N}} <: AbstractArray{T, N}
+    parent::A
+    function reinterpret(::Type{T}, a::A) where {T,N,S,A<:AbstractArray{S, N}}
+        function throwbits(::Type{S}, ::Type{T}, ::Type{U}) where {S,T,U}
+            @_noinline_meta
+            throw(ArgumentError("cannot reinterpret `$(S)` `$(T)`, type `$(U)` is not a bits type"))
+        end
+        function throwsize0(::Type{S}, ::Type{T})
+            @_noinline_meta
+            throw(ArgumentError("cannot reinterpret a zero-dimensional `$(S)` array to `$(T)` which is of a different size"))
+        end
+        function thrownonint(::Type{S}, ::Type{T}, dim)
+            @_noinline_meta
+            throw(ArgumentError("""
+                cannot reinterpret an `$(S)` array to `$(T)` whose first dimension has size `$(dim)`.
+                The resulting array would have non-integral first dimension.
+            """))
+        end
+        isbits(T) || throwbits(S, T, T)
+        isbits(S) || throwbits(S, T, S)
+        (N != 0 || sizeof(T) == sizeof(S)) || throwsize0(S, T)
+        if N != 0 && sizeof(S) != sizeof(T)
+            dim = size(a)[1]
+            rem(dim*sizeof(S),sizeof(T)) == 0 || thrownonint(S, T, dim)
+        end
+        new{T, N, S, A}(a)
+    end
+end
+
+parent(a::ReinterpretArray) = a.parent
+
+eltype(a::ReinterpretArray{T}) where {T} = T
+function size(a::ReinterpretArray{T,N,S} where {N}) where {T,S}
+    psize = size(a.parent)
+    size1 = div(psize[1]*sizeof(S), sizeof(T))
+    tuple(size1, tail(psize)...)
+end
+
+unsafe_convert(::Type{Ptr{T}}, a::ReinterpretArray{T,N,S} where N) where {T,S} = Ptr{T}(unsafe_convert(Ptr{S},a.parent))
+
+@inline @propagate_inbounds getindex(a::ReinterpretArray{T,0}) where {T} = reinterpret(T, a.parent[])
+@inline @propagate_inbounds getindex(a::ReinterpretArray) = a[1]
+
+@inline @propagate_inbounds function getindex(a::ReinterpretArray{T,N,S}, inds::Vararg{Int, N}) where {T,N,S}
+    if sizeof(T) == sizeof(S)
+        return reinterpret(T, a.parent[inds...])
```

Comment:
0.6:
```julia
julia> a = [(0x01,0x02,0x03,0x04)]
1-element Array{NTuple{4,UInt8},1}:
 (0x01, 0x02, 0x03, 0x04)

julia> reinterpret(Float32, a)
1-element Array{Float32,1}:
 1.53999f-36
```

This PR:
```julia
julia> a = [(0x01,0x02,0x03,0x04)]
1-element Array{NTuple{4,UInt8},1}:
 (0x01, 0x02, 0x03, 0x04)

julia> reinterpret(Float32, a)
1-element reinterpret(Float32, ::Array{NTuple{4,UInt8},1}):
Error showing value of type Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}:
ERROR: bitcast: expected primitive type value for second argument
Stacktrace:
 [1] getindex at ./reinterpretarray.jl:51 [inlined]
 [2] _getindex at ./abstractarray.jl:967 [inlined]
 [3] getindex(::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::Int64, ::Int64) at ./abstractarray.jl:926
 [4] isassigned(::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::Int64, ::Int64, ::Vararg{Int64,N} where N) at ./abstractarray.jl:258
 [5] alignment(::IOContext{Base.Terminals.TTYTerminal}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::Base.OneTo{Int64}, ::Base.OneTo{Int64}, ::Int64, ::Int64, ::Int64) at ./show.jl:1574
 [6] print_matrix(::IOContext{Base.Terminals.TTYTerminal}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::String, ::String, ::String, ::String, ::String, ::String, ::Int64, ::Int64) at ./show.jl:1703
 [7] print_matrix(::IOContext{Base.Terminals.TTYTerminal}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::String, ::String, ::String) at ./show.jl:1676
 [8] #showarray#284(::Bool, ::Function, ::IOContext{Base.Terminals.TTYTerminal}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}, ::Bool) at ./show.jl:2033
 [9] show(::IOContext{Base.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}) at ./replutil.jl:139
 [10] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}) at ./repl/REPL.jl:125
 [11] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}) at ./repl/REPL.jl:128
 [12] display(::Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}) at ./multimedia.jl:277
 [13] (::getfield(Base, Symbol("#inner#4")){Array{Any,1},typeof(display),Tuple{Base.ReinterpretArray{Float32,1,NTuple{4,UInt8},Array{NTuple{4,UInt8},1}}}})() at ./essentials.jl:669
 [14] print_response(::Base.Terminals.TTYTerminal, ::Any, ::Void, ::Bool, ::Bool, ::Void) at ./repl/REPL.jl:146
 [15] print_response(::Base.REPL.LineEditREPL, ::Any, ::Void, ::Bool, ::Bool) at ./repl/REPL.jl:132
 [16] (::getfield(Base.REPL, Symbol("#do_respond#17")){Bool,getfield(Base.REPL, Symbol("##27#37")){Base.REPL.LineEditREPL,Base.REPL.REPLHistoryProvider},Base.REPL.LineEditREPL,Base.LineEdit.Prompt})(::Base.LineEdit.MIState, ::Base.GenericIOBuffer{Array{UInt8,1}}, ::Bool) at ./repl/REPL.jl:674
```

Reply:
Agreed, this branch is only valid for primitive datatypes (`countfields(T + S) == 0`)